### PR TITLE
[WIP] Datatype I/O fixes.

### PR DESCRIPTION
**Pull request by jmchilton**
Fixes all sorts of issues, misses some other ones. More tests though.

### CODE REVIEW

File: lib/galaxy/datatypes/sniff.py [hunk 1]
```diff
@@ -103,28 +103,89 @@ def stream_to_file(stream, suffix='', prefix='', dir=None, text=False, **kwd):
...
+
+                    # We have a block... maybe at the end of the file.
+                    partial_line = True
+                    fp.write(line)
+    except UnicodeDecodeError:
+        fd, temp_name = tempfile.mkstemp(prefix=tmp_prefix, dir=tmp_dir)
         i = 0
-    else:
-        i += 1
+        with io.open(fd, mode="wb") as fp:
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/sniff.py:137]: Is there a case where only doing this would lead wrong results ?

Comment (jmchilton) [reply lib/galaxy/datatypes/sniff.py:137]: https://github.com/techtonik/dos2unix/blob/master/dos2unix.py#L18

Hmm... I don't have confidence to say but I doubt it. This dos2unix Python function does not worry about encoding.

**Comment by jmchilton** [comment]
xref https://github.com/chardet/chardet/blob/master/chardet/__init__.py

Hmmm... the new 1.imzml file is detected as binary by the sniff peek stuff so upload fails. Fun, ugh.

File: test/unit/datatypes/test_sniff.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    rval = convert_newlines_sep2tabs(tf.name, tmp_prefix="gxtest", tmp_dir=tempfile.gettempdir())
+    assert rval == (2, None), rval
+    assert expected == open(tf.name).read()
+
+
+def test_convert_newlines_non_utf():
+    fname = get_test_fname("dosimzml")
+    rval = convert_newlines(fname, tmp_prefix="gxtest", tmp_dir=tempfile.gettempdir(), in_place=False)
+    new_file = rval[1]
+    assert open(new_file, "rb").read() == open(get_test_fname("1.imzml"), "rb").read()
```
Comment (mvdbeek) [inline test/unit/datatypes/test_sniff.py]:
```suggestion
    assert open(new_file, "rb").read() == open(get_test_fname("1imzml"), "rb").read()
```

File: test/unit/datatypes/test_sniff.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import tempfile
```
Comment (mvdbeek) [inline test/unit/datatypes/test_sniff.py:1]:
```suggestion
import io
import tempfile
```

File: test/unit/datatypes/test_sniff.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def assert_converts_to_1234_sep2tabs(content, line_ending="\n"):
+    print("\r" in content)
+    tf = tempfile.NamedTemporaryFile(delete=False, mode='w')
+    tf.write(content)
+    tf.close()
+    rval = sep2tabs(tf.name, tmp_prefix="gxtest", tmp_dir=tempfile.gettempdir())
+    assert rval == (2, None), rval
+    assert '1\t2%s3\t4%s' % (line_ending, line_ending) == open(tf.name).read()
```
Comment (mvdbeek) [inline test/unit/datatypes/test_sniff.py:18]:
```suggestion
    assert '1\t2%s3\t4%s' % (line_ending, line_ending) == io.open(tf.name, newline='').read()
```

**Pull request merged**